lib/Sema/TypeCheckDeclPrimary.cpp

```diff
@@ -3656,3 +3663,103 @@ void TypeChecker::checkParameterList(ParameterList *params,
     diagnoseDuplicateDecls(*params);
   }
 }
+
+ArrayRef<Decl *>
+ExpandMacroExpansionDeclRequest::evaluate(Evaluator &evaluator,
+                                          MacroExpansionDecl *MED) const {
+  auto &ctx = MED->getASTContext();
+  auto *dc = MED->getDeclContext();
+  auto foundMacros = TypeChecker::lookupMacros(
+      MED->getDeclContext(), MED->getMacro(),
+      MED->getLoc(), MacroContext::FreestandingDeclaration);
+  if (foundMacros.empty()) {
+    MED->diagnose(diag::macro_undefined, MED->getMacro().getBaseIdentifier())
+        .highlight(MED->getMacroLoc().getSourceRange());
+    return {};
+  }
+  // Resolve macro candidates.
+  MacroDecl *macro;
+  // Non-call declaration macros cannot be overloaded.
+  auto *args = MED->getArgs();
+  if (!args) {
+    if (foundMacros.size() > 1) {
+      MED->diagnose(diag::ambiguous_decl_ref, MED->getMacro())
+          .highlight(MED->getMacroLoc().getSourceRange());
+      for (auto *candidate : foundMacros)
+        candidate->diagnose(diag::found_candidate);
+      return {};
+    }
+    macro = foundMacros.front();
+  }
+    // Call-like macros need to be resolved.
+  else {
+    using namespace constraints;
+    ConstraintSystem cs(dc, ConstraintSystemOptions());
```

Comment:
This looks a whole lot like CSGen for `MacroExpansionExpr`. Can we unify the implementations?

Reply:
I think we can pull it out to `TypeCheckMacros.h`. I can do it in a separate patch.